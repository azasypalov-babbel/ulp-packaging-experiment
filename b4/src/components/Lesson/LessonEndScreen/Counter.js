import React from 'react';
import PropTypes from 'prop-types';

export const animateValue = (value, onValueChange, onCompleted) => {
  if (value == 0) {
    onCompleted();
    onValueChange(0);
    return;
  }

  let current = 0;
  const stepTime = Math.floor(900 / value);
  const timer = setInterval(() => {
    current += 1;
    onValueChange(current);
    if (current == value) {
      onCompleted();
      clearInterval(timer);
    }
  }, stepTime);

  return timer;
};

class Counter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: 0
    };

    this.updateState = this.updateState.bind(this);
  }

  componentWillUnmount() {
    clearInterval(this.timer);
  }

  componentDidMount() {
    this.timer = animateValue(this.props.value, this.updateState, this.props.onCounterEnd);
  }

  updateState(newValue) {
    this.setState({ value: newValue });
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    clearInterval(this.timer);
    this.timer = animateValue(nextProps.value, this.updateState, this.props.onCounterEnd);
  }

  render() {
    return (
      <span>{this.state.value}</span>
    );
  }
}

Counter.propTypes = {
  value: PropTypes.number.isRequired,
  onCounterEnd: PropTypes.func
};

Counter.defaultProps = {
  onCounterEnd: () => {}
};

export default Counter;
