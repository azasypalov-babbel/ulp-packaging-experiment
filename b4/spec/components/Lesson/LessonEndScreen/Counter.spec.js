import React from 'react';
import Counter, { animateValue } from '../../../../src/components/Lesson/LessonEndScreen/Counter';
import { shallow } from 'enzyme';

const defaultProps = {
  value: 10
};

jest.useFakeTimers();

describe('<Counter />', () => {
  test('it renders', () => {
    const component = shallow(
      <Counter {...defaultProps} />
    );

    // Fast-forward until all timers have been executed
    jest.runAllTimers();
    // Forces a re-render since shallow won't do it by itself on changing state
    component.update();

    expect(component).toMatchSnapshot();
  });

  describe('#animateValue', () => {
    test('it calls onCompleted when count reaches value. (zero value)', () => {
      const onCountEnd = jest.fn();
      animateValue(0, () => {}, onCountEnd);
      expect(onCountEnd).toHaveBeenCalled();
    });

    test('it calls onCompleted when count reaches value.', () => {
      const onCountEnd = jest.fn();
      animateValue(1, () => {}, onCountEnd);
      jest.runOnlyPendingTimers();
      expect(onCountEnd).toHaveBeenCalled();
    });
  });
});
