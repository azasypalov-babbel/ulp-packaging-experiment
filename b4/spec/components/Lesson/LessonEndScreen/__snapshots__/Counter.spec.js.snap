// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> it renders 1`] = `
<span>
  10
</span>
`;
